analytics/endpoint_data_objects.go

```diff
@@ -0,0 +1,91 @@
+package analytics
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"time"
+)
+
+type RequestType string
+
+const (
+	COOKIE_SYNC RequestType = "/cookie_sync"
+	AUCTION     RequestType = "/openrtb2/auction"
+	SETUID      RequestType = "/set_uid"
+	AMP         RequestType = "/openrtb2/amp"
+)
+
+//Loggable object of a transaction at /openrtb2/auction endpoint
+type AuctionObject struct {
+	Type      RequestType
+	Status    int
+	Error     []error
+	Request   *openrtb.BidRequest
+	Response  *openrtb.BidResponse
+	UserAgent string
+}
+
+//Loggable object of a transaction at /openrtb2/amp endpoint
+type AmpObject struct {
+	Type            RequestType
+	Status          int
+	Error           []error
+	Request         *openrtb.BidRequest
+	AuctionResponse *openrtb.BidResponse
+	AmpResponse     map[string]string
+	UserAgent       string
+	Origin          string
+}
+
+//Loggable object of a transaction at /setuid
+type SetUIDObject struct {
+	Type     RequestType
+	Status   int
+	Bidder   string
+	UID      string
+	Usersync bool
+	Expiry   time.Duration
+	Error    []error
+	Success  bool
+}
+
+//Loggable object of a transaction at /cookie_sync
+type CookieSyncObject struct {
+	Type    RequestType
+	Status  int
+	Error   []error
+	Bidders string
```

Comment:
String - since I couldn't use a structured object owing to cyclic dependencies. Will have to pull out `cookieSyncResponse` from the _main_ package or `PBSBidders` from _pbs_ package in order to make the `Bidders` field structured.

Reply:
Once we can drop the legacy format we can probably restructure this better. :)

Reply:
Yup!!

Reply:
hmmm.... I'm actually pretty troubled by this. There are good reasons to prevent circular dependencies, and changing types to circumvent them is a very slippery slope.

I played around with this a bit locally and got the following import cycle:

```
package github.com/prebid/prebid-server
        imports github.com/prebid/prebid-server/adapters
        imports github.com/prebid/prebid-server/pbs
        imports github.com/prebid/prebid-server/analytics
        imports github.com/prebid/prebid-server/pbs
```

More specifically, this makes `analytics` depend on`pbs.PBSBidder`, while `pbs.SetUID()` also depends on `analytics`. And the reason you want to use `PBSBidder` is because it's used by `main.CookieSync()`.

It'd be one thing if this were all legacy code... but it's actually not. Both `/cookie_sync` and `/setuid` are "live", and we have no plans to delete either one anytime soon.

Stepping back a bit: the general trouble is that `analytics` wants to use the same _contract classes_ as the `endpoints` do. So an `endpoint` Impl must be able to "see" `analytics`, while `analytics` must be able to "see" the endpoint contract classes.

In OpenRTB, this works just fine. The contract types live in `mxmCherry/openrtb` or `openrtb_ext`, while the `endpoints` live in `endpoints/openrtb2`. In the legacy code, they're not cleanly separated. `PBSBidder` is the _contract type_ for `main.CookieSync()`, and it lives in _the same module_ as the `pbs.SetUID()` endpoint impl.

There are a few ways we could untangle this. In an ideal world, we'd move `SetUID()` and `CookieSync()` to `endpoints`... but I think that's a bit overkill for the issue you're hitting here.

I propose we:

- Move `pbs.UserSyncInfo` into the `usersync` module. 
- Make a new struct for `CookieSync()` with _just_ those 3 properties that it needs, and put that in the `usersync` module too.
- Make `CookieSync()` use this struct instead of `PBSBidder`.
- Make `analytics` use this new struct instead of `string`.

Conceptually, this makes the `usersync` module analogous to `openrtb_ext`. Unless I'm missing a dependency somewhere, `main` would depend on `analytics`, `pbs`, and `usersync`. `analytics` would depend on `usersync` only, and you'd get the structured type that you want here.

It may also be worth doing these moves in a separate PR? Your choice there... smaller changesets are usually easier to review and merge, and help minimize conflicts... but I'm not sure which is better.